pyomo/core/base/component.py

```diff
@@ -846,8 +850,32 @@ def index(self):
         - for some unknown reason - this instance does not belong
         to the parent component's index set.
         """
-        if self.parent_component() is None:
+        parent = self.parent_component()
+        if parent is None:
             return None
+        idx = self._index
+        # ComponentList objects implement _data as a list rather than a dict so
+        # we have to account for both here
+        if isinstance(parent._data, dict):
```

Comment:
Would a simpler test just be to:
```python
if parent[self._index] is not self:
    raise DeveloperError(...)
return self._index
```

Reply:
You get a KeyError if self._index is NOTSET. So, it serves the purpose, but the error message isn't very good.

Reply:
Ok, but that should be easy to correct, right?
```python
if self._index is not NOTSET and parent[self._index] is not self:
    raise DeveloperError(...)
return self._index
```
That does raise the question: should `NOTSET` raise an exception (which is the current implementation), or should it just be returned?  I think I am leaning toward the latter (and is proposed in this comment).

Reply:
OK, you convinced me, but just documenting here for my future self: We decided that `NOTSET` should not raise an exception, and neither should not having a parent: both cases should just return `self._index`. The only case that is necessarily an error is if the `_index` on the ComponentData is mapped to something different in the parent's `_data` dictionary. (Though more restrictive checks might be useful for debugging.)